core/scroll/scroll.service.js

```
@@ -0,0 +1,163 @@
+			}
+		}));
+
+		this.using(windowListener.on('resize', () => {
+			this.setElementsState(this.getMarkup());
+		}));
+	}
+
+	getMarkup() {
+		return this.table.view.markup;
+	}
+
+	canScroll(e) {
+		const table = this.table;
+
+		if (e.clientY < (table.bottom - offset) &&
+			e.clientY > (table.top + offset) &&
+			e.clientX > (table.left + offset) &&
+			e.clientX < (table.right - offset)) {
+			this.stop();
+		}
+	}
+
+	scroll(e) {
+		if (!this.interval) {
+			const area = this.detectArea(e);
+
+			switch (area) {
+				case 'top': {
+					const interval = this.doScroll('top', '-');
+					this.interval = interval();
+					break;
+				}
+				case 'bottom': {
+					const interval = this.doScroll('bottom', '+');
+					this.interval = interval();
+					break;
+				}
+				case 'left': {
+					const interval = this.doScroll('left', '-');
+					this.interval = interval();
+					break;
+				}
+				case 'right': {
+					const interval = this.doScroll('right', '+');
+					this.interval = interval();
+				}
+			}
+		}
+	}
+
+	doScroll(direction, modifier) {
+		const scrollVelocity = modifier === '+' ? velocity : velocity * -1;
+		const scrollState = this.model.scroll;
+		let scrolledToEnd;
+
+		switch (direction) {
+			case 'top': {
+				scrolledToEnd = () => this.isScrolledToEnd('top');
+				direction = 'top';
+				break;
+			}
+			case 'bottom': {
+				scrolledToEnd = () => this.isScrolledToEnd('bottom');
+				direction = 'top';
+				break;
+			}
+			case 'left': {
+				scrolledToEnd = () => this.isScrolledToEnd('left');
+				direction = 'left';
+				break;
+			}
+			case 'right': {
+				scrolledToEnd = () => this.isScrolledToEnd('right');
+				direction = 'left';
+			}
+		}
+
+		return () => setInterval(() => {
+			if(!scrolledToEnd()) {
+				const origin = scrollState()[direction];
+				scrollState({[direction]: origin + scrollVelocity});
+			}
+		}, 50);
+	}
+
+	isScrolledToEnd(direction) {
+		const body = this.body;
+
+		switch (direction) {
+			case 'top': {
```

default should raise an exception